Userland/Services/ContentAccessServer/ConnectionFromClient.cpp

```diff
@@ -0,0 +1,289 @@
+/*
+ * Copyright (c) 2021, timmot <tiwwot@protonmail.com>
+ * Copyright (c) 2021-2023, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/NumberFormat.h>
+#include <ContentAccessServer/ConnectionFromClient.h>
+#include <LibCore/File.h>
+#include <LibCore/Stream.h>
+#include <LibGUI/Application.h>
+#include <LibGUI/ConnectionToWindowServer.h>
+#include <LibGUI/FilePicker.h>
+#include <LibGUI/Label.h>
+#include <LibGUI/MessageBox.h>
+#include <LibGUI/Progressbar.h>
+#include <LibProtocol/Request.h>
+
+namespace ContentAccessServer {
+
+static HashMap<int, NonnullRefPtr<ConnectionFromClient>> s_connections;
+
+ConnectionFromClient::ConnectionFromClient(NonnullOwnPtr<Core::Stream::LocalSocket> socket)
+    : IPC::ConnectionFromClient<ContentAccessClientEndpoint, ContentAccessServerEndpoint>(*this, move(socket), 1)
+{
+    s_connections.set(1, *this);
+}
+
+void ConnectionFromClient::die()
+{
+    s_connections.remove(client_id());
+    GUI::Application::the()->quit();
+}
+
+RefPtr<GUI::Window> ConnectionFromClient::create_dummy_child_window(i32 window_server_client_id, i32 parent_window_id)
+{
+    auto window = GUI::Window::construct();
+    window->set_opacity(0);
+    window->set_frameless(true);
+    window->set_window_mode(GUI::WindowMode::Passive);
+    auto rect = GUI::ConnectionToWindowServer::the().get_window_rect_from_client(window_server_client_id, parent_window_id);
+    window->set_rect(rect);
+    window->show();
+    GUI::ConnectionToWindowServer::the().set_window_parent_from_client(window_server_client_id, parent_window_id, window->window_id());
+
+    return window;
+}
+
+ConnectionFromClient::ProgressObject ConnectionFromClient::create_download_progress_window(i32 window_server_client_id, i32 parent_window_id)
+{
+    auto main_window = create_dummy_child_window(window_server_client_id, parent_window_id);
+    auto dialog = GUI::Window::construct(main_window);
+    dialog->set_title("Downloading Remote File..."sv);
+    dialog->set_rect(Gfx::IntRect { 0, 0, 400, 100 });
+    dialog->center_within(*main_window);
+    auto frame = MUST(dialog->set_main_widget<GUI::Widget>());
+    MUST(frame->load_from_gml(R"gmlgmlgml(
```

Comment:
i thought it maybe would be better to go with `TRY`/`ErrorOr` instead (if possible), as this is a new piece of code…? (pushing `MUST` as high as possible, that would probably L238? 🤔)

Reply:
if we can't load the GML, there's no point in doing anything at all, might as well crash right where it failed imo.

Reply:
thereʼs something else here to be `TRIED`… but sure thing! 👍